activerecord/lib/active_record/relation/predicate_builder/association_query_value.rb

```diff
@@ -37,6 +40,22 @@ def primary_key
           associated_table.join_primary_key
         end
 
+        def primary_type
+          associated_table.join_primary_type
+        end
+
+        def polymorphic_name
+          associated_table.polymorphic_name_association
+        end
+
+        def select_values?
+          value.select_values.present?
+        end
+
+        def polymorphic_clause?
+          primary_type && !value.where_values_hash.has_key?(primary_type)
```

Comment:
If we want to use `where` we can follow the pattern for `select` and only add the condition if it's not present.